src/replication.c

```diff
@@ -419,12 +424,6 @@ void feedReplicationBuffer(char *s, size_t len) {
         if (add_new_block) {
             createReplicationBacklogIndex(listLast(server.repl_buffer_blocks));
         }
-        /* Try to trim replication backlog since replication backlog may exceed
-         * our setting when we add replication stream. Note that it is important to
-         * try to trim at least one node since in the common case this is where one
-         * new backlog node is added and one should be removed. See also comments
-         * in freeMemoryGetNotCountedMemory for details. */
-        incrementalTrimReplicationBacklog(REPL_BACKLOG_TRIM_BLOCKS_PER_CALL);
```

Comment:
this call being inside a loop (both `while(len > 0)` and the `argc` in the caller means that we have a loop that consumes memory (allocates replBufBlock), and it no longer frees memory.
so moving it outside could mean that we blow up the backlog, or at the very least not re-use existing memory (though the allocator's tcache).

maybe instead, we can keep running it inside the loop, but with a condition to do so only for cases were a new node was allocated (i.e. `add_new_block`)?
or alternatively, check why it is so expensive (seem to me that in most cases it'll exit rather quickly).

i do agree that when adding a plain command with several small arguments, there's no reason to try to trim the backlog per argument, but on a huge argument, we probably want to keep it inside the loop.

Reply:
For why it's so expensive, you're right that it usually exits relatively quickly on this condition:  https://github.com/BCathcart/redis/blob/optimize-repl-backlog-trimming/src/replication.c#L281-L282.

There doesn't appear to anything in particular that's expensive, but the repeated calls add up. Profiling shows `incrementalTrimReplicationBacklog()` makes up about 17% of the time spent in `replicationFeedSlaves()` for the 100 byte SET command workload (and was 3.8% of the total redis-server process CPU usage).

```
 |--13.82%--replicationFeedSlaves
|          |
 |          |--7.55%--feedReplicationBuffer.part.10
|          |          |
|          |          |--2.19%--incrementalTrimReplicationBacklog.constprop.32
|          |          |
|          |          |--0.66%--listNext
|          |          |
|          |           --0.14%--listRewind
|          |
|          |--1.76%--__memcpy_avx_unaligned_erms
|          |
|          |--0.90%--ll2string
|          |
|          |--0.65%--listNext
|          |
|          |--0.43%--stringObjectLen
|          |
|          |--0.35%--listRewind
|          |
|          |--0.20%--incrementalTrimReplicationBacklog.constprop.32
|          |
|           --0.10%--prepareClientToWrite
```

I understand the concern with large arguments. I was motivated to do it this way by the chain replication case which only trims after feeding the entire command from the query buffer to the replication backlog, whatever size it may be.

I assume we'll see a similar performance boost with trimming inside `feedReplicationBuffer()` only when we add a new block. Doing so could lead to an extra `PROTO_REPLY_CHUNK_BYTES` amount of data in the replication backlog from what we have right now if we allocate a new block right before hitting the condition on which we trim, but I presume that's an acceptable amount. Should I go ahead and update the PR to trim when adding a new block?

Reply:
yes, let's do that.